File: app/routes/messages.py
"""Direct messaging between users + block list.

DMs are private; only sender and recipient can read. Admin can list/delete
any DM for moderation.
"""
from __future__ import annotations

import re
from datetime import datetime
from functools import wraps

from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required, get_jwt, get_jwt_identity
from sqlalchemy import or_, and_, func

from app import db, limiter
from app.models import BlockedUser, DirectMessage, User
from app.routes.user_auth import current_user, user_required

messages_bp = Blueprint("messages", __name__)


def _admin_required(fn):
    @wraps(fn)
    @jwt_required()
    def wrapper(*args, **kwargs):
        if get_jwt().get("role") == "user":
            return jsonify({"error": "Admin access required"}), 403
        return fn(*args, **kwargs)
    return wrapper


_URL_RE = re.compile(r"https?://\S+", re.IGNORECASE)


def _sanitize(text: str, limit: int = 2000) -> str:
    text = (text or "").strip()
    if len(text) > limit:
        text = text[:limit]
    text = _URL_RE.sub("[link removed]", text)
    return text


# ============================ User: DMs ====================================

@messages_bp.route("/threads", methods=["GET"])
@user_required
def list_threads():
    """Return one row per conversation partner with last message + unread count."""
    me = current_user()

    # Latest message per (other_user_id) ordered by created_at desc
    rows = DirectMessage.query.filter(
        or_(DirectMessage.sender_id == me.id, DirectMessage.recipient_id == me.id)
    ).order_by(DirectMessage.created_at.desc()).limit(500).all()

    threads = {}
    for m in rows:
        other_id = m.recipient_id if m.sender_id == me.id else m.sender_id
        if other_id not in threads:
            threads[other_id] = {
                "other_user": (m.recipient if m.sender_id == me.id else m.sender),
                "last_message": m,
                "unread": 0,
            }
        if m.recipient_id == me.id and not m.is_read:
            threads[other_id]["unread"] += 1

    out = []
    for other_id, t in threads.items():
        ou = t["other_user"]
        out.append({
            "other_user": ou.public_dict() if ou else None,
            "last_message": t["last_message"].to_dict(viewer_id=me.id),
            "unread": t["unread"],
        })
    out.sort(key=lambda x: x["last_message"]["created_at"] or "", reverse=True)
    return jsonify(out), 200


@messages_bp.route("/with/<int:other_id>", methods=["GET"])
@user_required
def get_thread(other_id: int):
    me = current_user()
    if other_id == me.id:
        return jsonify({"error": "Cannot DM yourself"}), 400
    other = User.query.get_or_404(other_id)

    rows = DirectMessage.query.filter(
        or_(
            and_(DirectMessage.sender_id == me.id, DirectMessage.recipient_id == other.id),
            and_(DirectMessage.sender_id == other.id, DirectMessage.recipient_id == me.id),
        )
    ).order_by(DirectMessage.created_at.asc()).limit(500).all()

    # Mark recipient-side messages as read
    unread = [m for m in rows if m.recipient_id == me.id and not m.is_read]
    for m in unread:
        m.is_read = True
    if unread:
        db.session.commit()

    return jsonify({
        "other_user": other.public_dict(),
        "messages": [m.to_dict(viewer_id=me.id) for m in rows],
    }), 200


@messages_bp.route("/with/<int:other_id>", methods=["POST"])
@user_required
@limiter.limit("30 per minute")
def send_dm(other_id: int):
    me = current_user()
    if other_id == me.id:
        return jsonify({"error": "Cannot DM yourself"}), 400
    other = User.query.get_or_404(other_id)

    if other.is_banned:
        return jsonify({"error": "User is unavailable"}), 403
    if me.chat_muted:
        return jsonify({"error": "You are muted"}), 403
    if not other.notif_dm:
        return jsonify({"error": "Recipient has DMs disabled"}), 403

    blocked = BlockedUser.query.filter(
        or_(
            and_(BlockedUser.blocker_id == me.id, BlockedUser.blocked_id == other.id),
            and_(BlockedUser.blocker_id == other.id, BlockedUser.blocked_id == me.id),
        )
    ).first()
    if blocked:
        return jsonify({"error": "Cannot send message"}), 403

    data = request.get_json() or {}
    content = _sanitize(data.get("content") or "", 2000)
    if not content:
        return jsonify({"error": "Message is empty"}), 400

    m = DirectMessage(sender_id=me.id, recipient_id=other.id, content=content)
    db.session.add(m)
    db.session.commit()
    return jsonify(m.to_dict(viewer_id=me.id)), 201


@messages_bp.route("/unread-count", methods=["GET"])
@user_required
def unread_count():
    me = current_user()
    n = DirectMessage.query.filter_by(recipient_id=me.id, is_read=False, is_deleted=False).count()
    return jsonify({"unread": n}), 200


# ============================ User: blocks ==================================

@messages_bp.route("/blocks", methods=["GET"])
@user_required
def list_blocks():
    me = current_user()
    rows = BlockedUser.query.filter_by(blocker_id=me.id).all()
    users = User.query.filter(User.id.in_([r.blocked_id for r in rows])).all() if rows else []
    return jsonify([u.public_dict() for u in users]), 200


@messages_bp.route("/blocks/<int:other_id>", methods=["POST"])
@user_required
def block_user(other_id: int):
    me = current_user()
    if other_id == me.id:
        return jsonify({"error": "Cannot block yourself"}), 400
    if not User.query.get(other_id):
        return jsonify({"error": "User not found"}), 404
    if BlockedUser.query.filter_by(blocker_id=me.id, blocked_id=other_id).first():
        return jsonify({"message": "Already blocked"}), 200
    db.session.add(BlockedUser(blocker_id=me.id, blocked_id=other_id))
    db.session.commit()
    return jsonify({"message": "Blocked"}), 201


@messages_bp.route("/blocks/<int:other_id>", methods=["DELETE"])
@user_required
def unblock_user(other_id: int):
    me = current_user()
    row = BlockedUser.query.filter_by(blocker_id=me.id, blocked_id=other_id).first()
    if row:
        db.session.delete(row)
        db.session.commit()
    return jsonify({"message": "Unblocked"}), 200


# ============================ Admin moderation ==============================

@messages_bp.route("/admin/dms", methods=["GET"])
@_admin_required
def admin_list_dms():
    try:
        page = max(1, int(request.args.get("page", 1)))
        per_page = min(100, max(5, int(request.args.get("per_page", 50))))
    except ValueError:
        page, per_page = 1, 50

    q = DirectMessage.query
    only = (request.args.get("only") or "all").lower()
    if only == "active":
        q = q.filter_by(is_deleted=False)
    elif only == "deleted":
        q = q.filter_by(is_deleted=True)

    search = (request.args.get("search") or "").strip()
    if search:
        q = q.filter(DirectMessage.content.like(f"%{search}%"))

    pagination = q.order_by(DirectMessage.created_at.desc())\
                  .paginate(page=page, per_page=per_page, error_out=False)
    return jsonify({
        "messages": [m.to_dict() for m in pagination.items],
        "total": pagination.total,
        "page": pagination.page,
        "pages": pagination.pages,
    }), 200


@messages_bp.route("/admin/dms/<int:msg_id>", methods=["DELETE"])
@_admin_required
def admin_delete_dm(msg_id: int):
    m = DirectMessage.query.get_or_404(msg_id)
    m.is_deleted = True
    try:
        m.deleted_by_admin_id = int(get_jwt_identity())
    except (TypeError, ValueError):
        pass
    db.session.commit()
    return jsonify(m.to_dict()), 200

File: server/src/routes/messages.ts
import { Router } from 'express';
import { Types } from 'mongoose';
import { z } from 'zod';
import { BlockedUser, DirectMessage, User, conversationKey } from '../models/index.js';
import { asyncHandler } from '../lib/async-handler.js';
import { HttpError } from '../lib/http-error.js';
import { objectId, parseBody, parseQuery } from '../lib/validate.js';
import { serializeDirectMessage, serializeUser } from '../lib/serializers.js';
import { sanitizeChat } from '../lib/sanitize.js';
import { currentUser, requireUser } from '../middleware/auth.js';
import { dmLimiter } from '../middleware/rate-limit.js';
import { publishDirectMessage } from '../realtime/publish.js';

export const messagesRouter: Router = Router();

const DM_MAX = 2000;

/** Is either side blocking the other? */
async function blockedBetween(a: Types.ObjectId | string, b: Types.ObjectId | string) {
  return BlockedUser.exists({
    $or: [
      { blockerId: a, blockedId: b },
      { blockerId: b, blockedId: a },
    ],
  });
}

// ------------------------------------------------------------------ threads

/**
 * One row per conversation, with the last message and an unread count.
 *
 * Grouped in the database. The Flask version pulled the last 500 messages into
 * memory and bucketed them there, which silently dropped older conversations
 * and under-counted unread messages once an account got busy.
 */
messagesRouter.get(
  '/threads',
  requireUser,
  asyncHandler(async (req, res) => {
    const me = currentUser(req)._id;

    const rows = await DirectMessage.aggregate<{
      _id: string;
      lastMessage: Record<string, unknown>;
      unread: number;
    }>([
      { $match: { $or: [{ senderId: me }, { recipientId: me }] } },
      { $sort: { createdAt: -1 } },
      {
        $group: {
          _id: '$pairKey',
          lastMessage: { $first: '$$ROOT' },
          unread: {
            $sum: {
              $cond: [
                {
                  $and: [
                    { $eq: ['$recipientId', me] },
                    { $eq: ['$readAt', null] },
                    { $eq: ['$isDeleted', false] },
                  ],
                },
                1,
                0,
              ],
            },
          },
        },
      },
      { $sort: { 'lastMessage.createdAt': -1 } },
      { $limit: 100 },
    ]);

    const otherIds = rows.map((row) =>
      String(row.lastMessage.senderId) === String(me)
        ? row.lastMessage.recipientId
        : row.lastMessage.senderId,
    );

    // One extra query for every participant, rather than one per thread.
    const others = await User.find({ _id: { $in: otherIds } }).populate(
      'linkedPlayerId',
      'nameEn title',
    );
    const byId = new Map(others.map((user) => [String(user._id), user]));

    res.json(
      rows.map((row, index) => {
        const other = byId.get(String(otherIds[index]));
        return {
          other_user: other ? serializeUser(other) : null,
          last_message: serializeDirectMessage(
            row.lastMessage as Parameters<typeof serializeDirectMessage>[0],
            String(me),
          ),
          unread: row.unread,
        };
      }),
    );
  }),
);

messagesRouter.get(
  '/unread-count',
  requireUser,
  asyncHandler(async (req, res) => {
    const unread = await DirectMessage.countDocuments({
      recipientId: currentUser(req)._id,
      readAt: null,
      isDeleted: false,
    });
    res.json({ unread });
  }),
);

// ----------------------------------------------------------- conversation

messagesRouter.get(
  '/with/:userId',
  requireUser,
  asyncHandler(async (req, res) => {
    const otherId = objectId.parse(req.params.userId);
    const me = currentUser(req);

    if (otherId === String(me._id)) throw HttpError.badRequest('You cannot message yourself');

    const other = await User.findById(otherId).populate('linkedPlayerId', 'nameEn title');
    if (!other) throw HttpError.notFound('Player not found');

    // A block hides the conversation in both directions, rather than only
    // stopping new messages while the history stayed readable.
    if (await blockedBetween(me._id, other._id)) {
      throw HttpError.forbidden('This conversation is unavailable');
    }

    const { limit } = parseQuery(
      z.object({ limit: z.coerce.number().int().min(1).max(200).default(100) }),
      req,
    );

    const pairKey = conversationKey(String(me._id), otherId);
    const messages = await DirectMessage.find({ pairKey })
      .sort({ createdAt: -1 })
      .limit(limit)
      .populate('senderId', 'username displayName')
      .populate('recipientId', 'username displayName');

    messages.reverse();

    // Mark everything addressed to me as read, in one write.
    await DirectMessage.updateMany(
      { pairKey, recipientId: me._id, readAt: null },
      { $set: { readAt: new Date() } },
    );

    res.json({
      other_user: serializeUser(other),
      messages: messages.map((message) => serializeDirectMessage(message, String(me._id))),
    });
  }),
);

const sendSchema = z.object({
  content: z.string().min(1, 'Message is empty').max(4000),
});

messagesRouter.post(
  '/with/:userId',
  requireUser,
  dmLimiter,
  asyncHandler(async (req, res) => {
    const otherId = objectId.parse(req.params.userId);
    const { content } = parseBody(sendSchema, req);
    const me = currentUser(req);

    if (otherId === String(me._id)) throw HttpError.badRequest('You cannot message yourself');

    const other = await User.findById(otherId);
    if (!other) throw HttpError.notFound('Player not found');
    if (other.isBanned) throw HttpError.forbidden('This player is unavailable');
    if (me.chatMuted) throw HttpError.forbidden('You are muted');
    if (!other.notifDm) throw HttpError.forbidden('This player has direct messages turned off');
    if (await blockedBetween(me._id, other._id)) {
      throw HttpError.forbidden('You cannot message this player');
    }

    // Links are kept in direct messages: unlike in-game chat, a DM between two
    // people who chose to talk is a reasonable place to share one.
    const clean = sanitizeChat(content, DM_MAX, false);
    if (!clean) throw HttpError.badRequest('Message is empty');

    const message = await DirectMessage.create({
      senderId: me._id,
      recipientId: other._id,
      content: clean,
      pairKey: conversationKey(String(me._id), otherId),
    });
    await message.populate('senderId', 'username displayName');
    await message.populate('recipientId', 'username displayName');

    const payload = serializeDirectMessage(message, String(me._id));
    // Delivered live to the recipient's personal channel.
    publishDirectMessage(otherId, serializeDirectMessage(message, otherId));

    res.status(201).json(payload);
  }),
);

// ------------------------------------------------------------------ blocks

messagesRouter.get(
  '/blocks',
  requireUser,
  asyncHandler(async (req, res) => {
    const blocks = await BlockedUser.find({ blockerId: currentUser(req)._id }).select('blockedId');
    const users = await User.find({ _id: { $in: blocks.map((block) => block.blockedId) } });
    res.json(users.map(serializeUser));
  }),
);

messagesRouter.post(
  '/blocks/:userId',
  requireUser,
  asyncHandler(async (req, res) => {
    const otherId = objectId.parse(req.params.userId);
    const me = currentUser(req);

    if (otherId === String(me._id)) throw HttpError.badRequest('You cannot block yourself');
    if (!(await User.exists({ _id: otherId }))) throw HttpError.notFound('Player not found');

    // Upsert, so blocking twice is not an error.
    await BlockedUser.updateOne(
      { blockerId: me._id, blockedId: otherId },
      { $setOnInsert: { blockerId: me._id, blockedId: otherId } },
      { upsert: true },
    );

    res.status(201).json({ message: 'Blocked' });
  }),
);

messagesRouter.delete(
  '/blocks/:userId',
  requireUser,
  asyncHandler(async (req, res) => {
    const otherId = objectId.parse(req.params.userId);
    await BlockedUser.deleteOne({ blockerId: currentUser(req)._id, blockedId: otherId });
    res.json({ message: 'Unblocked' });
  }),
);
